File: 30.substring-with-concatenation-of-all-words.cpp
/*
 * @lc app=leetcode id=30 lang=cpp
 *
 * [30] Substring with Concatenation of All Words
 */

// @lc code=start
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int m = words[0].size();
        int n = s.size();
        int min = 0;
        vector<int> res;
        unordered_map<string, int> wordCount;
        for (auto word : words) {
            ++wordCount[word];
        }
        for (int i = 0; i < m; i++) {
            int count = 0;
            unordered_map<string, int> map;
            for (int l = i, r = i; r + m <= n; r += m) {
                string str = s.substr(r, m);
                ++map[str];
                if (map[str] <= wordCount[str]) {
                    ++count;
                }
                if (count == words.size()) {
                    string str = s.substr(l, m);
                    while (map[str] > wordCount[str]) {
                        l += m;
                        --map[str];
                        str = s.substr(l, m);
                    }
                    int len = r - l;
                    if (len == (words.size() - 1) * m) {
                        res.push_back(l);
                    }
                    --map[str];
                    --count;
                    l += m;
                }
            }
        }

        return res;
    }
};
// @lc code=end

File: 30.substring-with-concatenation-of-all-words.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*
 * @lc app=leetcode id=30 lang=java
 *
 * [30] Substring with Concatenation of All Words
 */

// @lc code=start
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (words.length == 0) {
            return result;
        }
        Map<String, Integer> had = new HashMap<>();
        for (String word : words) {
            had.put(word, had.getOrDefault(word, 0) + 1);
        }
        int m = s.length(), n = words[0].length(), len = words.length;
        Map<String, Integer> found = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int start = i, end = i, count = 0; end + n <= m; end += n) {
                String w = s.substring(end, end + n);
                found.put(w, found.getOrDefault(w, 0) + 1);
                if (had.containsKey(w) && found.get(w).intValue() <= had.get(w).intValue()) {
                    count++;
                }
                if (count == len) {
                    String sw = s.substring(start, start + n);
                    while (!had.containsKey(sw) || found.get(sw).intValue() > had.get(sw).intValue()) {
                        start += n;
                        found.put(sw, found.get(sw) - 1);
                        sw = s.substring(start, start + n);
                    }
                    if (end - start + n == n * len) {
                        result.add(start);
                    }
                    start += n;
                    found.put(sw, found.get(sw) - 1);
                    count--;
                }
            }
            found.clear();
        }
        return result;
    }
}
// @lc code=end

